stack/src/lib.rs

```diff
@@ -872,4 +872,122 @@ mod tests {
+            Item::Value(MyValue::Bytes(hash_secret.as_ref().to_vec())),
+            Item::Operator(MyOperator::Equal),
+            Item::Operator(MyOperator::Verify),
+            Item::Value(MyValue::Bytes(pk_2.pkh().bytes().to_vec())),
+            Item::Operator(MyOperator::CheckSig),
+            Item::Operator(MyOperator::Verify),
+            Item::Operator(MyOperator::EndIf),
+        ];
+        assert!(!execute_script(s, &ScriptContext { block_timestamp: 0 }));
+
+        // 2 can spend with secret
+        let s = vec![
+            // Witness script
+            Item::Value(MyValue::Signature(ks_2.to_pb_bytes().unwrap())),
+            Item::Value(MyValue::Bytes(secret)),
+            Item::Value(MyValue::Boolean(false)),
+            // Redeem script
+            Item::Operator(MyOperator::If),
+            Item::Value(MyValue::Integer(10_000)),
+            Item::Operator(MyOperator::CheckTimeLock),
+            Item::Operator(MyOperator::Verify),
+            Item::Value(MyValue::Bytes(pk_1.pkh().bytes().to_vec())),
+            Item::Operator(MyOperator::CheckSig),
+            Item::Operator(MyOperator::Verify),
+            Item::Operator(MyOperator::Else),
+            Item::Operator(MyOperator::Sha256),
+            Item::Value(MyValue::Bytes(hash_secret.as_ref().to_vec())),
+            Item::Operator(MyOperator::Equal),
+            Item::Operator(MyOperator::Verify),
+            Item::Value(MyValue::Bytes(pk_2.pkh().bytes().to_vec())),
+            Item::Operator(MyOperator::CheckSig),
+            Item::Operator(MyOperator::Verify),
+            Item::Operator(MyOperator::EndIf),
+        ];
+        assert!(execute_script(s, &ScriptContext { block_timestamp: 0 }));
+
+        // 2 cannot spend with a wrong secret
+        let s = vec![
+            // Witness script
+            Item::Value(MyValue::Signature(ks_2.to_pb_bytes().unwrap())),
+            Item::Value(MyValue::Bytes(vec![0, 0, 0, 0])),
+            Item::Value(MyValue::Boolean(false)),
+            // Redeem script
+            Item::Operator(MyOperator::If),
+            Item::Value(MyValue::Integer(10_000)),
+            Item::Operator(MyOperator::CheckTimeLock),
+            Item::Operator(MyOperator::Verify),
+            Item::Value(MyValue::Bytes(pk_1.pkh().bytes().to_vec())),
+            Item::Operator(MyOperator::CheckSig),
+            Item::Operator(MyOperator::Verify),
+            Item::Operator(MyOperator::Else),
+            Item::Operator(MyOperator::Sha256),
+            Item::Value(MyValue::Bytes(hash_secret.as_ref().to_vec())),
+            Item::Operator(MyOperator::Equal),
+            Item::Operator(MyOperator::Verify),
+            Item::Value(MyValue::Bytes(pk_2.pkh().bytes().to_vec())),
+            Item::Operator(MyOperator::CheckSig),
+            Item::Operator(MyOperator::Verify),
+            Item::Operator(MyOperator::EndIf),
+        ];
+        assert!(!execute_script(s, &ScriptContext { block_timestamp: 0 }));
+    }
```

Comment:
Should we check that 2 cannot spend after timelock even if the right secret is provided? :thinking:

Reply:
I think that, ideally, they shouldn't. It is a sort of assumption for an atomic swap that only 1 party can spend at a time.